test/erlumts_nas_security_tests.erl

```diff
@@ -50,26 +52,104 @@ ts33401_test_sets() ->
       hexstream_to_binary(<<"5cb72c6edc878f1566e10253afc364c9fa540d914db94cbee275d0917ca6af0d77acb4ef3bbe1a722b2ef5bd1d4b8e2aa5024ec1388a201e7bce7920aec615895f763a5564dcc4c482a2ee1d8bfecc4498eca83fbb75f9ab530e0dafbede2fa5895b82991b6277c529e0f2529d7f79606be96706296dedfa9d7412b616958cb563c678c02825c30d0aee77c4c146d2765412421a808d13cec819694c75ad572e9b973d948b81a9337c3b2a17192e22c2069f7ed1162af44cdea817603665e807ce40c8e0dd9d6394dc6e31153fe1955c47afb51f2617ee0c5e3b8ef1ad7574ed343edc2743cc94c990e1f1fd264253c178dea739c0befeebcd9f9b76d49c1015c9fecf50e53b8b5204dbcd3eed863855dabcdcc94b31e318021568855c8b9e52a981957a112827f978ba960f1447911b317b5511fbcc7fb13ac153db74251117e4861eb9e83bffffc4eb7755579038e57924b1f78b3e1ad90bab2a07871b72db5eef96c334044966db0c37cafd1a89e5646a3580eb6465f121dce9cb88d85b96cf23ccccd4280767bee8eeb23d8652461db6493103003baf89f5e18261ea43c84a92ebffffe4909dc46c5192f825f770600b9602c557b5f8b431a79d45977dd9c41b863da9e142e90020cfd074d6927b7ab3b6725d1a6f3f98b9c9daa8982aff06782800">>)}].
 
 encrypt_eea2_128_test() ->
-    ok = encrypt_tests(ts33401_test_sets()).
+    ok = encrypt_tests(ts33401_ciphering_test_sets()).
 
 encrypt_tests([]) ->
     ok;
 encrypt_tests([TestCase|Rest]) ->
-    {Key, Count, Bearer, Dir, Length, PlainText, CipherText} = TestCase,
-    Out = erlumts_nas_security:encrypt(eea2_128, Key, Count, Bearer, Dir, PlainText, Length),
+    {Key, Count, Bearer, Direction, Length, PlainText, CipherText} = TestCase,
+    Out = erlumts_nas_security:encrypt(eea2_128, Key, Count, Bearer, Direction, PlainText, Length),
     ?assertEqual(CipherText, Out),
     encrypt_tests(Rest).
 
 decrypt_eea2_128_test() ->
-    ok = decrypt_tests(ts33401_test_sets()).
+    ok = decrypt_tests(ts33401_ciphering_test_sets()).
 
 decrypt_tests([]) ->
     ok;
 decrypt_tests([TestCase|Rest]) ->
-    {Key, Count, Bearer, Dir, Length, PlainText, CipherText} = TestCase,
-    Out = erlumts_nas_security:decrypt(eea2_128, Key, Count, Bearer, Dir, CipherText, Length),
+    {Key, Count, Bearer, Direction, Length, PlainText, CipherText} = TestCase,
+    Out = erlumts_nas_security:decrypt(eea2_128, Key, Count, Bearer, Direction, CipherText, Length),
     ?assertEqual(PlainText, Out),
     decrypt_tests(Rest).
 
+%% Integrity protection
+
+ts33401_integrity_test_sets() ->
+    %% TS 33.401 C.2.1 - C.2.8
+    %% format: {Key, Count, Bearer, Direction, Length, Message, MAC}
+    %% The message in most of the test cases 3GPP provides are not
+    %% aligned to the 8-bit boundary, which we don't currently support.
+    [%{hexstream_to_binary(<<"2bd6459f82c5b300952c49104881ff48">>),
+     % hexstream_to_binary(<<"38a6f056">>),
+     % 16#18,
+     % 0,
+     % 58,
+     % hexstream_to_binary(<<"3332346263393840">>),
+     % hexstream_to_binary(<<"118c6eb8">>)},
```

Comment:
lots of commented test code, to use or not to use?

Reply:
It should pass once the problem with the weird length message is fixed. Removed them for now, as I'm not sure when we really work on that fix.